File: thread_test/src/spin.py
#!/usr/bin/env python3

#   A simple python ROS node with 2 subscribers
#   with rospy.spin() thread spinning model.
#   Author : Tony Jacob

import rospy
from std_msgs.msg import String, Int16
import time

class ThreadChecker:
    def __init__(self):
        rospy.init_node("python_threadchecker")
        rospy.Subscriber("/string",String, callback=self.StringCB)
        rospy.Subscriber("/int",Int16,callback=self.IntCB)
        rospy.loginfo("[Python] Node Started")

    def StringCB(self,msg):
        timer = 8
        rospy.loginfo(f"[Python] Received string, Sleeping for {timer} seconds")
        time.sleep(timer)
        rospy.loginfo("[Python] String thread opened from sleep")
    
    def IntCB(self,msg):
        timer = 10
        rospy.loginfo(f"[Python] Received Int, Sleeping for {timer} seconds")
        time.sleep(timer)
        rospy.loginfo("[Python] Int thread opened from sleep")

ThreadChecker()
rospy.spin()

File: thread_test/src/spin.cpp
/*A simple C++ ROS Node with 2 subscriber
callbacks with a ros::spin() thread spinning model

Author: Tony Jacob*/

#include "ros/ros.h"
#include "ros/console.h"
#include "std_msgs/String.h"
#include "std_msgs/Int16.h"
#include <time.h>

class ThreadChecker {
    private:
        ros::Subscriber stringsub;
        ros::Subscriber intsub;
    
    public:
        ThreadChecker(ros::NodeHandle *nh){
            ROS_INFO("[Spin] Node Started");
            stringsub = nh->subscribe("/string",10, &ThreadChecker::callback_string, this );
            intsub = nh->subscribe("/int",10,&ThreadChecker::callback_int, this);
        }

        void callback_string(const std_msgs::String &msg){
            int timer = 8;
            ROS_INFO("[Spin] Received string, Sleeping for %d seconds", timer);
            sleep(timer);
            ROS_INFO("[Spin] String thread opened from sleep");
        }

        void callback_int(const std_msgs::Int16 &msg){
            int timer = 10;
            ROS_INFO("[Spin] Received Int, Sleeping for %d seconds", timer);
            sleep(timer);
            ROS_INFO("[Spin] Int thread opened from sleep");
        }

};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "threadchecker");
    ros::NodeHandle nh;
    ThreadChecker tc = ThreadChecker(&nh);
    ros::spin();
}